RenderStateNotationParser/src/RenderStateNotationParserImpl.cpp

```diff
@@ -302,43 +305,126 @@ RenderStateNotationParserImpl::RenderStateNotationParserImpl(IReferenceCounters*
+                        *Name = ResourceDesc.Name;
+                }
+                else
+                {
+                    throw nlohmann::json::type_error::create(JsonTypeError, std::string("type must be object or string, but is ") + Json.type_name(), Json);
+                }
+            };
+
+            if (Json.contains("Defaults"))
             {
-                ShaderCreateInfo ResourceDesc = {};
-                Deserialize(Shader, ResourceDesc, *m_pAllocator);
-                VERIFY_EXPR(ResourceDesc.Desc.Name != nullptr);
-                m_ShaderNames.emplace(HashMapStringKey{ResourceDesc.Desc.Name, false}, StaticCast<Uint32>(m_Shaders.size()));
-                m_Shaders.push_back(ResourceDesc);
+                auto const& Default = Json["Defaults"];
+
+                if (Default.contains("Shader"))
+                    Deserialize(Default["Shader"], DefaultShader, *m_pAllocator);
+
+                if (Default.contains("Pipeline"))
+                    Deserialize(Default["Pipeline"], DefaultPipeline, *m_pAllocator, Callbacks);
             }
 
+            for (auto const& Shader : Json["Shaders"])
+                Callbacks.ShaderCallback(Shader, SHADER_TYPE_UNKNOWN, nullptr, *m_pAllocator);
+
             for (auto const& RenderPass : Json["RenderPasses"])
-            {
-                RenderPassDesc ResourceDesc = {};
-                Deserialize(RenderPass, ResourceDesc, *m_pAllocator);
-                VERIFY_EXPR(ResourceDesc.Name != nullptr);
-                m_RenderPassNames.emplace(HashMapStringKey{ResourceDesc.Name, false}, StaticCast<Uint32>(m_RenderPasses.size()));
-                m_RenderPasses.push_back(ResourceDesc);
-            }
+                Callbacks.RenderPassCallback(RenderPass, nullptr, *m_pAllocator);
 
             for (auto const& Signature : Json["ResourceSignatures"])
-            {
-                PipelineResourceSignatureDesc ResourceDesc = {};
-                Deserialize(Signature, ResourceDesc, *m_pAllocator);
-                VERIFY_EXPR(ResourceDesc.Name != nullptr);
-                m_ResourceSignatureNames.emplace(HashMapStringKey{ResourceDesc.Name, false}, StaticCast<Uint32>(m_ResourceSignatures.size()));
-                m_ResourceSignatures.push_back(ResourceDesc);
-            }
+                Callbacks.ResourceSignatureCallback(Signature, nullptr, *m_pAllocator);
+
+
 
             for (auto const& Pipeline : Json["Pipelines"])
             {
-                auto& PipelineType = Pipeline.at("PSODesc").at("PipelineType");
+                auto AssignPipelineState = [](PipelineStateNotation& RHS, PipelineStateNotation& LHS, PIPELINE_TYPE Type) {
+                    RHS.PSODesc                     = LHS.PSODesc;
+                    RHS.Flags                       = LHS.Flags;
+                    RHS.ResourceSignaturesNameCount = LHS.ResourceSignaturesNameCount;
+                    RHS.ppResourceSignatureNames    = LHS.ppResourceSignatureNames;
```

Comment:
Why assign one-by-one? `RHS = LHS` is shorter and safer (in case new members are added to the `PipelineStateNotation`)

Reply:
Besides, RHS and LHS are mixed up.

Reply:
Source should be `const &`